### PR TITLE
Add ReadVariableOp gradient to C++

### LINKED ISSUE #46114: C++ Gradient for ReadVariableOp
**System information**
- TensorFlow version (you are using): `master`
- Are you willing to contribute it: Yes

**Describe the feature and the current behavior/state.**

Currently, `ReadVariableOp`'s gradient is implemented in Python, making using resource variables for training from the C++ API or other APIs based on it impossible.

**Will this change the current api? How?**

No.

**Who will benefit with this feature?**

Anyone using resource variables from the C++ API or APIs based on it.

### PR DESCRIPTION
**Pull request by rnett**
Fixes #46114.

Adds the ResourceVariableOp gradient to the C++ API and removes it from the Python API.

### CODE REVIEW

**Comment by rnett** [comment]
@suharshs I'm attempting to add the C++ gradient for `ReadVariableOp` following the instructions you laid out in [the /tensorflow/cc/gradients README](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/gradients/README.md), which matches past PRs such as #25467.  However, it does not appear to be registered, as seen in `//tensorflow/python/kernel_tests:cond_v2_test` and other tests.  It doesn't appear to be related to using a new file as it happens even if I move the new gradient to `array_grad.cc`.  Thoughts?

File: tensorflow/python/ops/resource_variable_ops.py [hunk 1]
```diff
@@ -2132,12 +2132,6 @@ def op(self):
     return self._parent_op
 
 
-@ops.RegisterGradient("ReadVariableOp")
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/python/ops/resource_variable_ops.py:2135]: Ah, that explains the errors.  Is there any other way to avoid the duplication?

[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/python/ops/resource_variable_ops.py:2135]: There is a current plan to use C++ gradients from Python, but they're over at https://github.com/tensorflow/tensorflow/tree/master/tensorflow/c/experimental/gradients

I don't see resource op gradients there yet. It's probably something we'll want eventually. (CC @saxenasaurabh)

Unfortunately there have been several "write our gradients in C++" efforts over the years, and this set is not compatible with eager execution. So since we're supporting eager execution in Python we won't be able to use these.

File: tensorflow/cc/gradients/resource_variable_grad_test.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  Scope scope_;
+};
+
+TEST_F(ResourceVariableGradTest, ReadVariableOpGrad) {
+  TensorShape shape({5, 2});
+  auto x = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
+  auto var = VarHandleOp(scope_, DT_FLOAT, {});
+  auto init = AssignVariableOp(scope_, var, x);
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc:58]: Agreed, I'm not (intentionally) testing that.  I'm attempting to work with the existing test harness which afaik requires me to use a placeholder as input, ideally I'd get a variable and read from it.  The assignment is the equivalent of `tf.Variable(x)` so that it's initialized.  

The test report 404s now, but if I remember right this test passed.  I'll fix the python side and see.

File: tensorflow/python/ops/resource_variable_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/python/ops/resource_variable_ops.py:2135]: Oh, this is indented for tf Java, which also supports eager execution, should I move this there (and to that format)?

[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/python/ops/resource_variable_ops.py:2135]: I'm not sure where Java gets its gradient definitions. Probably better to start with a unit test for the behavior you want.

[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/python/ops/resource_variable_ops.py:2135]: It uses the C API.  And yeah, it looks like I'll have to test it from that end as well.

**Comment by rnett** [comment]
@allenlavoie we (i.e. SIG JVM) are going to need to add a number of gradients to bring the Java API to parity, can you check my understanding of where things live (or tell me if there's a nice guide somewhere I missed)?  I can't test eager and function myself, since there's no C tape api and the `SymbolicGradient` doesn't have a gradient registered with how we're doing graph gradients (`TF_AddGradientsWithPrefix` in the C API).

* Functions are differentiated with the `SymbolicGradient` op, where gradients are registered with `REGISTER_OP_GRADIENT` and live in `core/ops` like [this](https://github.com/tensorflow/tensorflow/blob/c7e7f49228444f1a024fbd9266d60252009fee0a/tensorflow/core/ops/resource_variable_ops.cc#L161).

* Graph gradients, at least those done through the C API's `TF_AddGradientsWithPrefix`, are registered with `REGISTER_GRADIENT_OP` and live in `cc/gradients`.  I assume if there was a way to do this using `SymbolicGradient`/`REGISTER_OP_GRADIENT` you already would be?

* Eager gradients live in `c/experimental/gradients` and registered by adding them to tape's [`RegisterGradients`](https://github.com/tensorflow/tensorflow/blob/4e25cac495909450b42094f7f6c85d49bbed1c2f/tensorflow/python/framework/experimental/tape.cc#L30).

I don't suppose there's any attempts to unify some or all of these?

**Comment by allenlavoie** [comment]
I see. Yes, if you're using SymbolicGradient/TF_AddGradientsWithPrefix then cc/gradients sounds right for adding new definitions. SymbolicGradient isn't used much from Python anymore for function gradients, since there isn't really a way to get higher-order gradients with it. I'd like to put a replacement for function gradients in C (based on replaying the function) and use it from Python, but it's not something I'll get to soon. And similarly an eager GradientTape doesn't do higher-order gradients if the gradient functions it runs are opaque, which makes re-using the graph-building gradient definitions pretty hard; they need to be written using the same API the tape is watching.

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/c/experimental/gradients is the effort to unify Python/C++/eager/graph to use one set of gradients. The gradient definitions are written with an eager/graph agnostic C++ API (with a C option likely for ABI stability), and the plan is to use them from Python early (something we didn't manage for most of these other projects). I'm not sure it's far enough along to have a registry via static initializers; looks like the tests are still manually adding definitions to a registry. And there is some uncertainty as we figure out what kind of C++ API we want going forward.

File: tensorflow/python/ops/resource_variable_ops.py [hunk 3]
```diff
@@ -2131,10 +2131,10 @@ def op(self):
     """The op for this variable."""
     return self._parent_op
 
-
+
```
Comment (allenlavoie) [inline tensorflow/python/ops/resource_variable_ops.py:2134]: Can you revert the whitespace changes in this file?

File: tensorflow/cc/gradients/resource_variable_grad_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc:62]: I see. Yeah, I guess if you want to use the numeric checker you need to do something like this.

What I don't understand is why there is a gradient of `x` with respect to `y`; `var` with respect to `y` makes sense to me, but if assignment has a gradient too then there's a rabbit hole that I don't think is implemented correctly right now (the gradient only flowing to the second assignment at execution time, which isn't statically known).

**Comment by rnett** [comment]
> https://github.com/tensorflow/tensorflow/tree/master/tensorflow/c/experimental/gradients is the effort to unify Python/C++/eager/graph to use one set of gradients.

Ok, thanks, that's good to know.  I assume `TF_AddGradientsWithPrefix` will use them eventually?  And is there an issue or PR I can watch to keep up to date on progress with them?

Also related: I'm looking at adding `cc/gradients` gradients for `PartitionedCall` (and Stateful), is calling the `SymbolicGradient` op still the right way to do that or should I look at using `SymbolicGradientBuilder` (I know you said it's obsolete, but we don't have a c++ level alternative and it's still used as a fallback [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradients_util.py#L308-L308))

**Comment by allenlavoie** [comment]
Sure, fine to use it for now. I don't know of an issue tracking the eager/graph agnostic C++ gradients (@saxenasaurabh might).

SymbolicGradient is OK for first-order gradients of functions. Do you need functions and function gradients if there's no tape / eager-compatible gradient API?

**Comment by rnett** [comment]
> SymbolicGradient is OK for first-order gradients of functions. Do you need functions and function gradients if there's no tape / eager-compatible gradient API?

Yeah, we're supporting functions in graphs as well as eager mode.  Eventually we'll point people towards the eager API, but without gradients it's not functional yet.

File: tensorflow/cc/gradients/resource_variable_grad_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc:61]: I'd just manually take a gradient of the resource handle with respect to the output of the ReadVariableOp (or with respect to the Sum to reduce it to a scalar if it isn't already). It's just a ones tensor with the shape of the ReadVariableOp output, so you don't really need the gradient checker. (Or if you want to make it more interesting, you can do something after the read op; either way the read op is identity in the backward pass.)

At least, that's how the tape models variables in Python. There is a can of worms around the gradient of a scalar-shape resource handle being a non-scalar (shape/dtype of the ReadVariableOp output). There's a question of how much we want to do for these graph-only C++ gradients vs. doing it once with the eager/graph agnostic API.

[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc]: I didn't run into the scalar shape issue.  I don't think we need to do much with the C++ API, I just want this as a bandaid so I can start working on variables in the Java API.

**Comment by rnett** [comment]
Had to add an Identity to the gradient to prevent "fetching and feeding the same op" issues.  Test passes locally now, I expect everything to work.

File: tensorflow/cc/gradients/resource_variable_grad_test.cc [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#include "tensorflow/cc/gradients/grad_testutil.h"
+#include "tensorflow/core/framework/tensor_testutil.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+
+namespace tensorflow {
+namespace {
+
+using namespace ops;  // NOLINT(build/namespaces)
+
+class ResourceVariableGradTest : public ::testing::Test {
```
Comment (allenlavoie) [inline tensorflow/cc/gradients/resource_variable_grad_test.cc:34]: I think this test fixture can be deleted now, and you can just use TEST(ArbitrarySuiteName, TestName) instead of TEST_F

**Review by allenlavoie — APPROVED** [review]

File: tensorflow/cc/gradients/resource_variable_grad_test.cc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  auto dy = ops::Cast(scope_, ops::Const(scope_, 1.0, shape), DT_FLOAT);
+
+  OutputList dxs;
+  TF_ASSERT_OK(AddSymbolicGradients(scope_, {y}, {var}, {dy}, &dxs));
+
+
+  ClientSession::FeedType feed_list;
+  feed_list.insert({x, (float) 5});
+  feed_list.insert({dy, (float) 1});
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc:78]: No, but it's less confusing to have it all in one place.  I'll make it a placeholder.

**Review by allenlavoie — APPROVED** [review]

**Review by allenlavoie — APPROVED** [review]

**Comment by rthadur** [comment]
@rnett can you please check for sanity build failures ?

**Comment by rnett** [comment]
Should be fixed, it passes locally.

File: tensorflow/cc/gradients/resource_variable_grad_test.cc [hunk 6]
```diff
@@ -33,6 +33,7 @@ namespace {
 
 TEST(ResourceVariableGradTest, ReadVariableOpGrad) {
   TensorShape shape({});
+  auto scope_ = Scope::NewRootScope();
```
Comment (allenlavoie) [inline tensorflow/cc/gradients/resource_variable_grad_test.cc:36]: Why the trailing underscore? That's typically for class members, but this is a local variable.

Comment (rnett) [reply tensorflow/cc/gradients/resource_variable_grad_test.cc:36]: Since it was a class member and I didn't rename it.  I'll do so.

**Review by allenlavoie — APPROVED** [review]

**Comment by rthadur** [comment]
@rnett can you please fix sanity builds ?

**Review by allenlavoie — APPROVED** [review]

**Comment by rthadur** [comment]
@rnett we still see sanity build errors , can you please check ?

**Comment by rnett** [comment]
It keeps reporting new errors each time.  When I try running the [`tensorflow/tools/ci_build/ci_sanity.sh`](tensorflow/tools/ci_build/ci_sanity.sh) it hangs while looking for CUDA dirs in the configure script (which works fine normally).  Let me see if I ran just the buildifier.

**Comment by allenlavoie** [comment]
Is it complaining about the type of whitespace this time? I agree that's a pretty confusing error and it's probably easier to run buildifier. That looks like the only thing that's been failing.

```
FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
tensorflow/cc/BUILD:
555,557d554
<         "//tensorflow/core:test",
<         "//tensorflow/core:test_main",
<         "//tensorflow/core:testlib",
560a558,560
>         "//tensorflow/core:test",
>         "//tensorflow/core:test_main",
>         "//tensorflow/core:testlib",
exit status 1
```

**Comment by rnett** [comment]
Looks like it just changed the order again, the diff seems wrong wrt what it changed.  It passes `buildifier -v -mode=check BUILD` now though, locally.

**Review by allenlavoie — APPROVED** [review]

**Comment by allenlavoie** [comment]
Ah, that makes sense, "core" comes after "cc" alphabetically and the diff was just saying the lines stay in the same order but move down.

**Comment by allenlavoie** [comment]
Can you resolve the conflicts? (I'm guessing with your other change)

**Review by allenlavoie — APPROVED** [review]

**Comment by allenlavoie** [comment]
Can you run clang_tidy?

Specifically it wants the three C-style `(float)1` casts to be C++-style `static_cast<float>(1)`, although you can also just write `1.0f` (or `1f`?)

**Review by allenlavoie — APPROVED** [review]

**Review by allenlavoie — APPROVED** [review]

**Review by allenlavoie — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]
